source/shared/core_conn.cpp

```diff
@@ -150,34 +150,134 @@ sqlsrv_conn* core_sqlsrv_connect( _In_ sqlsrv_context& henv_cp, _In_ sqlsrv_cont
+            // Thus, we will simply throw an exception.
+            if( core_search_odbc_driver_unix( ODBC_DRIVER_17 ) ) {
+                conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION::ODBC_DRIVER_17 ];
+                r = core_odbc_connect( conn, conn_str, is_pooled );
+            }
+            else {
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_AE_DRIVER_REQUIRED, get_processor_arch()) {
+                    throw core::CoreException();
+                }
+            }
+        } // else driver_version not unknown
     }
-    else {
+    else { 
+        // column encryption NOT enabled
+        if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
+            // check if the ODBC driver actually exists, if not, throw an exception
+            CHECK_CUSTOM_ERROR( ! core_search_odbc_driver_unix( conn->driver_version ), conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND ) {
+                throw core::CoreException();
+            }
 
-        for ( std::size_t i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
-            is_missing_driver = false;
-            std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION(i) ];
-            r = core_odbc_connect( conn, conn_str_driver, is_missing_driver, is_pooled );
-            CHECK_CUSTOM_ERROR( is_missing_driver && ( i == DRIVER_VERSION::LAST ), conn, SQLSRV_ERROR_DRIVER_NOT_INSTALLED, get_processor_arch()) {
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+        }
+        else {
+            DRIVER_VERSION odbc_version = ODBC_DRIVER_UNKNOWN;
+            for( short i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
+                // skip ODBC 11 in a non-Windows environment -- only available in Red Hat / SUSE (preview)
+                // https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server#microsoft-odbc-driver-11-for-sql-server-on-linux
+                if (i == DRIVER_VERSION::ODBC_DRIVER_11)    
+                    continue;
+
+                if( core_search_odbc_driver_unix( DRIVER_VERSION( i ) ) ) {
+                    odbc_version = DRIVER_VERSION( i );
+                    break;
+                }
+            }
+            CHECK_CUSTOM_ERROR( odbc_version == ODBC_DRIVER_UNKNOWN, conn, SQLSRV_ERROR_DRIVER_NOT_INSTALLED, get_processor_arch()) {
                 throw core::CoreException();
             }
-            if ( !is_missing_driver) {
-                break;
+            std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION( odbc_version ) ];
+            r = core_odbc_connect( conn, conn_str_driver, is_pooled );
+        }  // else driver_version not unknown
+    } // else ce_option enabled
+#else
+    if( conn->ce_option.enabled ) {
+        // when AE is enabled, must use ODBC driver 17
+        if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
```

Comment:
this block is identical in both _WIN32 cases, move to before ifdef

Reply:
no it's not identical